### PR TITLE
Add svg elements and attributes

### PR DESCRIPTION
**Pull request by nileliao**

### CODE REVIEW

**Comment by randomdross** [comment]
I'd like to be really conservative about the elements / attributes added.  I saw you pulled out 'script' but there are plenty of other things that could be subtly dangerous.  I also suspect that support for anchors in SVG will need to be wired up in code (the same way anchors are supported today).  I haven't dug in deep yet, but this is my initial impression.

BTW, it's great to see jSanity getting some attention.  Are you planning on using it in production anywhere?

**Comment by nileliao** [comment]
The concern about the added elements/attributes is valid.
My new strategy would be using a list of elements/attributes that are already been validated as initial version and collecting telemetry to add more if necessary.

Yes, it will be used in production to do SVG sanitize.

**Comment by randomdross** [comment]
Ok, that sounds reasonable.  Maybe update the pull request with the smaller
list of elements / attributes and I'll do a bit more research to see if I
can find any in the list that could be problematic.

**Comment by cure53** [comment]
Hey all, feel free to check here for a safe list of SVG elements and attributes:

https://github.com/cure53/DOMPurify/tree/master/src

**Comment by nileliao** [comment]
@cure53 thanks

**Comment by nileliao** [comment]
Need to double check tag "use" and add IRI validation for attributes "requiredextensions", "requiredfeatures", "xlink:href" based on current list

File: jsanity-0.3.js [hunk 1]
```diff
@@ -774,7 +636,8 @@ jSanity = {};
             'http://': 1,
             'https://': 1,
             'ftp://': 1,
-            'mailto:': 1
+            'mailto:': 1,
+            '#': 1,
```
Comment (randomdross) [inline jsanity-0.3.js]: Maybe make this off by default, with a flag?  (similar to the yet-to-be-implemented allowRelativeURLs flag.)  I'd argue that it's not 100% guaranteed safe to allow sanitized markup to poke at various entry points on the current page without knowing the full URL.

Comment (nileliao) [reply jsanity-0.3.js]: Thanks, will revert this change. And after investigation, IRI sanitize is non-trivial to implement, I'll remove related stuff and revert related change for this pull.

**Comment by nileliao** [comment]
Comparing to 'https://github.com/cure53/DOMPurify/blob/master/src/tags.js':
Added:
feBlend, blend filter
Removed:
animatecolor
animatemotion
animatetransform
audio
canvas
video

Comparing to 'https://github.com/cure53/DOMPurify/blob/master/src/attrs.js':
Added:
amplitude: number
calcMode: keywords
clipPathUnits: keywords
dominant-baseline: keywords
exponent: number
filterUnits: keywords
intercept: number
limitingConeAngle: number
pointsAtX: number
pointsAtY: number
pointsAtZ: number
primitiveUnits: keywords
slope: number
startOffset: length
systemLanguage: language
tableValues: number list
unicode-bidi: keywords
Removed:
Attributes related to url/iri/link or not in internal list

**Comment by randomdross** [comment]
Code should probably be added so the image element supports externalContentCallback, because this element allows external images to be loaded similar to the HTML IMG tag.

Similarly, for anchors I think there needs to be support in allowLinks / linkClickCallback related code.

Less concretely, I'm slightly worried that something here might enable the overlay mitigation to be bypassed.  (In other words, that it would be possible to create markup that would scribble on top of other UI on the page, outside the region where sanitized content should appear.)  I haven't tried to actually substantiate this though -- it may not actually be possible.

I'm not trying to frustrate you away from getting this into the codebase, I'm just trying hold the bar super high.

One more (easy): Initialize unsupportedContentCallback to null (see dataAttributeCallback as an example).

If all this is reasonably addressed I don't anticipate any other roadblocks.  It will be great to get this SVG support.

**Comment by nileliao** [comment]
> Code should probably be added so the image element supports externalContentCallback, because this element allows external images to be loaded similar to the HTML IMG tag.
> Similarly, for anchors I think there needs to be support in allowLinks / linkClickCallback related code.

Even though `<a>, ` are allowed, `target, href, xlink:href, externalResourcesRequired` and other unsafe attributes are not allowed, which makes `<a>, ` actually useless. I plan not to support them in this version(mainly because non-trivial) and only support later based on usage. If you found it confusing, I can remove them from `knownElements`.
>Less concretely, I'm slightly worried that something here might enable the overlay mitigation to be bypassed. (In other words, that it would be possible to create markup that would scribble on top of other UI on the page, outside the region where sanitized content should appear.) I haven't tried to actually substantiate this though -- it may not actually be possible.

Agree. Also it might not be possible to know where **sanitized content should appear**. In this version, since links are disabled in SVG, I suppose it would be less risky even SVG content scribbles on top of other UI. It's definitely something worth revisit when supporting links for SVG.  
>I'm not trying to frustrate you away from getting this into the codebase, I'm just trying hold the bar super high.

Totally understand, thanks for the help :)

>One more (easy): Initialize unsupportedContentCallback to null (see dataAttributeCallback as an example).

unsupportedContentCallback is initialized to null in line 20, similar as dataAttributeCallback in line 56.  Am I missing something? 
```javascript
unsupportedContentCallback: null, // Callback for unsupported content, such as elements/attributes/styles
dataAttributeCallback: null, // Callback for handling of data-* attributes which are otherwise unsafe by default

**Comment by randomdross** [comment]
> Even though <a>,  are allowed, target, href, xlink:href, externalResourcesRequired and other unsafe attributes are not allowed, which makes <a>,  actually useless. I plan not to support them in this version(mainly because non-trivial) and only support later based on usage. If you found it confusing, I can remove them from knownElements.

That resolves my concern on this one.

> In this version, since links are disabled in SVG, I suppose it would be less risky even SVG content scribbles on top of other UI. 

Yes, it's a good point that without links it's harder to trick someone into doing something.  But even without links, if the attacker effectively can scribble over all the UI, they can at least in theory confuse the user into performing actions that are already possible within the app.  E.g.: Swapping "OK" and "Cancel" buttons.

> unsupportedContentCallback is initialized to null in line 20, similar as dataAttributeCallback in line 56. Am I missing something?

Ack, I missed this.  This is all good.

**Comment by nileliao** [comment]
>Yes, it's a good point that without links it's harder to trick someone into doing something. But even without links, if the attacker effectively can scribble over all the UI, they can at least in theory confuse the user into performing actions that are already possible within the app. E.g.: Swapping "OK" and "Cancel" buttons.

Good point, any suggestion on this?

**Comment by randomdross** [comment]
Ok, after playing around with it a bit and looking through the code, I feel like overlay actually won't be a problem.  First, there doesn't seem to be a way to simulate position:fixed or position:absolute with pure SVG (without CSS).  Without position:fixed or position:absolute, I think jSanity does a good job mitigating this already by forcing sanitized content into a fixed size box from which it can't escape:
https://github.com/Microsoft/JSanity/blob/master/jsanity-0.3.js#L1079

There is also code that explicitly kills position:fixed and position:absolute:
https://github.com/Microsoft/JSanity/blob/master/jsanity-0.3.js#L869

Let me know if you see anything I may have missed there though.

One more thing I realized...  The 'name' and 'class' SVG attributes in theory could allow for conflict with or abuse of existing content on the page.  However it looks like jSanity already handles both of these:
https://github.com/Microsoft/JSanity/blob/master/jsanity-0.3.js#L767
https://github.com/Microsoft/JSanity/blob/master/jsanity-0.3.js#L779
...and I think if these appear in SVG they will be handled by this code as well, even though it's in the SVG namespace.

So, I think we're good to go!  The only question I have now is how to do versioning.  I'm not sure why I had the version number in the filename, that seems to be wrong.  What do you think about introducing a version identifier in the code itself and just changing the filename to jSanity.js?

**Comment by nileliao** [comment]
>So, I think we're good to go! The only question I have now is how to do versioning. I'm not sure why I had the version number in the filename, that seems to be wrong. What do you think about introducing a version identifier in the code itself and just changing the filename to jSanity.js?

This is a good idea, less code change needed when upgrading version,

One remain work is to add safe svg presentation attributes to knownCSSProperties so that they can be kept in style after sanitize. I'll include them in this change and send a new iteration.

**Review by randomdross — APPROVED** [review]

**Pull request merged**